src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,406 @@
+
+        self.__next_step.submit(new_message)
+        self.__batch = None
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        self.__next_step.join(timeout)
+
+
+class ProducerResultFuture(NamedTuple):
+    message: Message[KafkaPayload]
+    future: Future
+
+
+class ProduceStep(ProcessingStep[MessageBatch]):
+    """
+    Step that produces to the snuba-metrics topic, collecting the futures returned by
+    the producer. Continously checks to see if futures are done, and once that's the case
+    can commit up to the last future that is done.
+    """
+
+    def __init__(self, commit_function: Callable[[Mapping[Partition, Position]], None]) -> None:
+        snuba_metrics = settings.KAFKA_TOPICS[settings.KAFKA_SNUBA_METRICS]
+        snuba_metrics_producer = KafkaProducer(
+            kafka_config.get_kafka_producer_cluster_options(snuba_metrics["cluster"]),
+        )
+        self.__producer = snuba_metrics_producer
+        self.__producer_topic = settings.KAFKA_TOPICS[settings.KAFKA_SNUBA_METRICS].get(
+            "topic", "snuba-metrics"
+        )
+        self.__commit_function = commit_function
+
+        self.__futures: Deque[ProducerResultFuture] = deque()
+        self.__closed = False
+
+        # TODO(meredith): make this an option to pass in
+        self.__max_buffer_size = 10000
+
+    def poll(self) -> None:
+        while self.__futures:
+            if not self.__futures[0].future.done():
+                break
+
+            result_future = self.__futures.popleft()
+            message, future = result_future
+
+            try:
+                future.result()
+            except Exception:
+                # TODO(meredith): log info for the different errors
+                # that could happen:
+                # * CancelledError (future was cancelled)
+                # * TimeoutError (future timedout)
+                # * Exception (the fucture call raised an exception)
+                raise
+            self.__commit_function({message.partition: Position(message.offset, message.timestamp)})
+
+    def submit(self, outer_message: Message[MessageBatch]) -> None:
+        assert not self.__closed
+
+        if len(self.__futures) >= self.__max_buffer_size:
+            raise MessageRejected
+
+        for message in outer_message.payload:
+            payload = message.payload
+            future = self.__producer.produce(
+                destination=Topic(self.__producer_topic),
+                payload=payload,
+            )
+            self.__futures.append(ProducerResultFuture(message, future))
+
+    def close(self) -> None:
+        self.__closed = True
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def join(self, timeout: Optional[float] = None) -> None:
```

Comment:
You need to close the producer at the end of this method or the consumer will hang when we try to stop it.